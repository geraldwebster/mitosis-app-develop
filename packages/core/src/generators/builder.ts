import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';

export type ToBuilderOptions = {
  prettier?: boolean;
};

export const blockToBuilder = (json: JSXLiteNode, options: ToBuilderOptions = {}) => {
  return null;
};

export const componentToBuilder = (
  componentJson: JSXLiteComponent,
  options: ToBuilderOptions = {},
) => {
  // Make a copy we can safely mutate, similar to babel's toolchain
  return null;
};
