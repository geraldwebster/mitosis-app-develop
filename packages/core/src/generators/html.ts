import { camelCase } from 'lodash';
import { format } from 'prettier/standalone';
import { getStateObjectString } from '../helpers/get-state-object-string';
import { collectCss } from '../helpers/collect-styles';
import { fastClone } from '../helpers/fast-clone';
import { selfClosingTags } from '../parsers/jsx';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';
import { dashCase } from '../helpers/dash-case';
import { isComponent } from '../helpers/is-component';
import { methodLiteralPrefix } from '../constants/method-literal-prefix';
import traverse from 'traverse';
import { isJsxLiteNode } from '../helpers/is-jsx-lite-node';
import { replaceIdentifiers } from '../helpers/replace-idenifiers';

type ToHtmlOptions = {
  prettier?: boolean;
};

type StringMap = { [key: string]: string };
type InternalToHtmlOptions = ToHtmlOptions & {
  onChangeJsById: StringMap;
  js: string;
};

const getGetters = (json: JSXLiteComponent) => {
  const getters: string[] = [];
  for (const key in json.state) {
    const value = json.state[key];
    if (typeof value === 'string') {
      if (value.startsWith(methodLiteralPrefix + 'get ')) {
        getters.push(key);
      }
    }
  }
  return getters;
};

const getForNames = (json: JSXLiteComponent) => {
  const names: string[] = [];
  traverse(json).forEach(function (item) {
    if (isJsxLiteNode(item)) {
      if (item.name === 'For') {
        names.push(item.bindings._forName as string);
      }
    }
  });
  return names;
};

const replaceForNameIdentifiers = (json: JSXLiteComponent) => {
  // TODO: cache this. by reference? lru?
  const forNames = getForNames(json);

  traverse(json).forEach((item) => {
    if (isJsxLiteNode(item)) {
      for (const key in item.bindings) {
        if (key === 'css' || key === '_forName') {
          continue;
        }
        const value = item.bindings[key];
        if (typeof value === 'string') {
          item.bindings[key] = replaceIdentifiers(
            value,
            forNames,
            (name) => `getContext(el, "${name}")`,
          ) as string;
        }
      }
    }
  });
};

const mappers: {
  [key: string]: (json: JSXLiteNode, options: InternalToHtmlOptions) => string;
} = {
  Fragment: (json, options) => {
    return json.children.map((item) => blockToHtml(item, options)).join('\n');
  },
};

const getId = (json: JSXLiteNode) =>
  `${dashCase(json.name)}-${Math.random().toString(26).slice(9)}`;

const addOnChangeJs = (
  id: string,
  options: InternalToHtmlOptions,
  code: string,
) => {
  if (!options.onChangeJsById[id]) {
    options.onChangeJsById[id] = '';
  }
  options.onChangeJsById[id] += code;
};

// TODO: spread support
const blockToHtml = (json: JSXLiteNode, options: InternalToHtmlOptions) => {
  const hasData = Object.keys(json.bindings).length;
  let elId = '';
  if (hasData) {
    elId = getId(json);
    json.properties['data-uid'] = elId;
  }

  if (mappers[json.name]) {
    return mappers[json.name](json, options);
  }

  if (json.properties._text) {
    return json.properties._text;
  }
  if (json.bindings._text) {
    addOnChangeJs(elId, options, `el.innerText = ${json.bindings._text};`);

    return `<span data-uid="${elId}"></span>`;
  }

  let str = '';

  if (json.name === 'For') {
    const itemName = json.bindings._forName;
    addOnChangeJs(
      elId,
      options,
      `
        el.innerHTML = '';
        var _array = ${json.bindings.each};
        var template = document.querySelector('[data-template="${elId}"]');
        for (var value of _array) {
          var tmp = document.createElement('span');
          tmp.innerHTML = template.innerHTML;
          Array.from(tmp.children).forEach(function (child) {
            contextMap.set(child, {
              ...contextMap.get(child),
              ${itemName}: value
            });
            el.appendChild(child);
          });
        }
      `,
    );
    // TODO: decide on how to handle this...
    str += `
      <span data-uid="${elId}"></span>
      <template data-template="${elId}">`;
    if (json.children) {
      str += json.children.map((item) => blockToHtml(item, options)).join('\n');
    }
    str += '</template>';
  } else if (json.name === 'Show') {
    addOnChangeJs(
      elId,
      options,
      `el.style.display = ${json.bindings.when} ? 'inline' : 'none'`,
    );

    str += `<span data-uid="${elId}">`;
    if (json.children) {
      str += json.children.map((item) => blockToHtml(item, options)).join('\n');
    }

    str += '</span>';
  } else {
    str += `<${json.name} `;

    // For now, spread is not supported
    // if (json.bindings._spread === '_spread') {
    //   str += `
    //       {% for _attr in ${json.bindings._spread} %}
    //         {{ _attr[0] }}="{{ _attr[1] }}"
    //       {% endfor %}
    //     `;
    // }

    for (const key in json.properties) {
      const value = json.properties[key];
      str += ` ${key}="${value}" `;
    }

    for (const key in json.bindings) {
      if (key === '_spread' || key === 'ref' || key === 'css') {
        continue;
      }
      const value = json.bindings[key] as string;
      // TODO: proper babel transform to replace. Util for this
      const useValue = value;

      if (key.startsWith('on')) {
        let event = key.replace('on', '').toLowerCase();
        if (!isComponent(json) && event === 'change') {
          event = 'input';
        }
        const fnName = camelCase(`on-${elId}-${event}`);
        options.js += `
          function ${fnName}(event) {
            ${useValue}
          }
        `;
        addOnChangeJs(
          elId,
          options,
          `
            el.removeEventListener('${event}', ${fnName});
            el.addEventListener('${event}', ${fnName});
          `,
        );
      } else {
        const useAttribute = key.includes('-');
        addOnChangeJs(
          elId,
          options,
          useAttribute
            ? `el.setAttribute(${key}, ${useValue})`
            : `el.${key} = ${useValue}`,
        );
      }
    }
    if (selfClosingTags.has(json.name)) {
      return str + ' />';
    }
    str += '>';
    if (json.children) {
      str += json.children.map((item) => blockToHtml(item, options)).join('\n');
    }

    str += `</${json.name}>`;
  }
  return str;
};

// TODO: props support via custom elements
export const componentToHtml = (
  componentJson: JSXLiteComponent,
  options: ToHtmlOptions = {},
) => {
  const useOptions: InternalToHtmlOptions = {
    ...options,
    onChangeJsById: {},
    js: '',
  };
  const json = fastClone(componentJson);
  replaceForNameIdentifiers(json);

  const hasState = Boolean(Object.keys(json.state).length);

  const css = collectCss(json);
  let str = json.children
    .map((item) => blockToHtml(item, useOptions))
    .join('\n');

  if (css.trim().length) {
    str += `<style>${css}</style>`;
  }

  if (hasState) {
    // TODO: collectJs helper for here and liquid
    str += `
      <script>
        var rawState = ${getStateObjectString(json)};

        var observers = [];
        var contextMap = new WeakMap();
        var state = new Proxy(rawState, {
          set(target, key, value, receiver) {
            Reflect.set(target, key, value, receiver);
            runObservers(key, value, receiver)
          },
        });

        onChange(function () {
          ${Object.keys(useOptions.onChangeJsById)
            .map((key) => {
              const value = useOptions.onChangeJsById[key];
              if (!value) {
                return '';
              }
              return `
              document.querySelectorAll("[data-uid='${key}']").forEach(function (el) {
                ${value}
              })
            `;
            })
            .join('\n\n')}
        })
        runObservers()

        ${useOptions.js}

        function runObservers(key, value, receiver) {
          observers.forEach(function (cb) {
            cb(key, value, receiver);
          })
        }
        function onChange(cb) {
          observers.push(cb);
        }
        function getContext(el, name) {
          var parent = el;
          do {
            var context = contextMap.get(parent);
            if (context && name in context) {
              return context[name];
            }
          } while (parent = parent.parentNode)
        }
      </script>
    `;
  }

  if (options.prettier !== false) {
    try {
      str = format(str, {
        parser: 'html',
        plugins: [
          // To support running in browsers
          require('prettier/parser-html'),
          require('prettier/parser-postcss'),
          require('prettier/parser-babel'),
        ],
      });
    } catch (err) {
      console.warn('Could not prettify', { string: str }, err);
    }
  }
  return str;
};
