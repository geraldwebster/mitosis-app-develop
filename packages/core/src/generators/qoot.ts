import dedent from 'dedent';
import { camelCase, kebabCase } from 'lodash';
import { format } from 'prettier/standalone';
import { isJsxLiteNode } from '../helpers/is-jsx-lite-node';
import traverse from 'traverse';
import { capitalize } from '../helpers/capitalize';
import { fastClone } from '../helpers/fast-clone';
import { renderPreComponent } from '../helpers/render-imports';
import { stripMetaProperties } from '../helpers/strip-meta-properties';
import { filterEmptyTextNodes } from '../helpers/filter-empty-text-nodes';
import {
  Plugin,
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../modules/plugins';
import { selfClosingTags } from '../parsers/jsx';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';
import { removeSurroundingBlock } from 'src/helpers/remove-surrounding-block';
import { getStateObjectString } from 'src/helpers/get-state-object-string';
import { stripStateAndPropsRefs } from 'src/helpers/strip-state-and-props-refs';
import { babelTransformExpression } from 'src/helpers/babel-transform';
import { NodePath, types } from '@babel/core';

function addMarkDirtyAfterSetInCode(
  code: string,
  options: InternalToQootOptions,
  useString = 'markDirty(this)',
) {
  return babelTransformExpression(code, {
    AssignmentExpression(
      path: babel.NodePath<babel.types.AssignmentExpression>,
    ) {
      const { node } = path;
      if (types.isMemberExpression(node.left)) {
        if (types.isIdentifier(node.left.object)) {
          // TODO: utillity to properly trace this reference to the beginning
          if (node.left.object.name === 'state') {
            // TODO: ultimately do updates by property, e.g. updateName()
            // that updates any attributes dependent on name, etc√ß
            let parent: NodePath<any> = path;

            // `_temp = ` assignments are created sometimes when we insertAfter
            // for simple expressions. this causes us to re-process the same expression
            // in an infinite loop
            while ((parent = parent.parentPath)) {
              if (
                types.isAssignmentExpression(parent.node) &&
                types.isIdentifier(parent.node.left) &&
                parent.node.left.name.startsWith('_temp')
              ) {
                return;
              }
            }

            path.insertAfter(
              types.callExpression(types.identifier(useString), []),
            );
          }
        }
      }
    },
  });
}

const processBinding = (binding: string, options: InternalToQootOptions) =>
  addMarkDirtyAfterSetInCode(
    stripStateAndPropsRefs(binding, {
      replaceWith: 'this.',
    }),
    options,
  );

const NODE_MAPPERS: {
  [key: string]: (json: JSXLiteNode, options: InternalToQootOptions) => string;
} = {
  Fragment(json, options) {
    return `<>${json.children
      .map((item) => blockToQoot(item, options))
      .join('\n')}</>`;
  },
  For(json, options) {
    return `{${processBinding(json.bindings.each as string, options)}.map(${
      json.bindings._forName
    } => (
      <>${json.children
        .filter(filterEmptyTextNodes)
        .map((item) => blockToQoot(item, options))
        .join('\n')}</>
    ))}`;
  },
  Show(json, options) {
    return `{Boolean(${processBinding(
      json.bindings.when as string,
      options,
    )}) && (
      <>${json.children
        .filter(filterEmptyTextNodes)
        .map((item) => blockToQoot(item, options))
        .join('\n')}</>
    )}`;
  },
};

const getId = (json: JSXLiteNode, options: InternalToQootOptions) => {
  const name = json.properties.$name
    ? camelCase(json.properties.$name)
    : /^h\d$/.test(json.name || '') // don't dashcase h1 into h-1
    ? json.name
    : camelCase(json.name || 'div');

  const newNameNum = (options.namesMap[name] || 0) + 1;
  options.namesMap[name] = newNameNum;
  return capitalize(`${name}${newNameNum === 1 ? '' : `${newNameNum}`}`);
};

// This should really be a preprocessor mapping the `class` attribute binding based on what other values have
// to make this more pluggable
const collectClassString = (json: JSXLiteNode): string | null => {
  const staticClasses: string[] = [];

  const hasStaticClasses = Boolean(staticClasses.length);
  if (json.properties.class) {
    staticClasses.push(json.properties.class);
    delete json.properties.class;
  }
  if (json.properties.className) {
    staticClasses.push(json.properties.className);
    delete json.properties.className;
  }

  const dynamicClasses: string[] = [];
  if (typeof json.bindings.class === 'string') {
    dynamicClasses.push(json.bindings.class as any);
    delete json.bindings.class;
  }
  if (typeof json.bindings.className === 'string') {
    dynamicClasses.push(json.bindings.className as any);
    delete json.bindings.className;
  }
  if (typeof json.bindings.className === 'string') {
    dynamicClasses.push(json.bindings.className as any);
    delete json.bindings.className;
  }
  if (typeof json.bindings.css === 'string') {
    dynamicClasses.push(`css(${json.bindings.css})`);
    delete json.bindings.css;
  }
  const staticClassesString = staticClasses.join(' ');

  const dynamicClassesString = dynamicClasses.join(" + ' ' + ");

  const hasDynamicClasses = Boolean(dynamicClasses.length);

  if (hasStaticClasses && !hasDynamicClasses) {
    return `"${staticClassesString}"`;
  }

  if (hasDynamicClasses && !hasStaticClasses) {
    return `{${dynamicClassesString}}`;
  }

  if (hasDynamicClasses && hasStaticClasses) {
    return `{"${staticClassesString} " + ${dynamicClassesString}}`;
  }

  return null;
};

type NumberRecord = { [key: string]: number };
type ToQootOptions = {
  prettier?: boolean;
  plugins?: Plugin[];
};
type InternalToQootOptions = {
  prettier?: boolean;
  plugins?: Plugin[];
  componentJson: JSXLiteComponent;
  namesMap: NumberRecord;
};
const blockToQoot = (json: JSXLiteNode, options: InternalToQootOptions) => {
  if (NODE_MAPPERS[json.name]) {
    return NODE_MAPPERS[json.name](json, options);
  }
  if (json.properties._text) {
    return json.properties._text;
  }
  if (json.bindings._text) {
    return `{${json.bindings._text}}`;
  }

  let str = '';

  const id = getId(json, options);
  json.meta.id = id;

  str += `<${json.name} `;

  const classString = collectClassString(json);
  if (classString) {
    str += ` class=${classString} `;
  }

  if (json.bindings._spread) {
    str += ` {...(${json.bindings._spread})} `;
  }

  for (const key in json.properties) {
    if (key.startsWith('_') || key.startsWith('$')) {
      continue;
    }
    const value = json.properties[key];
    str += ` ${key}="${value}" `;
  }
  for (const key in json.bindings) {
    const value = json.bindings[key] as string;
    if (key.startsWith('_') || key.startsWith('$')) {
      continue;
    }

    if (key.startsWith('on')) {
      // TODO: this transformation can be a IR transform middleware for rendering
      // through any framework
      // TODO: for now use _ instead of : until I find what voodoo magic allows
      // colons in attribute names in TSX
      const useKey = key.replace('on', 'on_').toLowerCase();
      const componentName = getComponentName(options.componentJson, options);
      // TODO: args
      // on:click={QRL`ui:/Item/toggle?toggleState=.target.checked`}
      str += ` ${useKey}={QRL\`ui:/${componentName}/on${id}${key.slice(2)}\`}`;
    } else {
      str += ` ${key}={${value}} `;
    }
  }
  if (selfClosingTags.has(json.name)) {
    return str + ' />';
  }
  str += '>';
  if (json.children) {
    str += json.children.map((item) => blockToQoot(item, options)).join('\n');
  }

  str += `</${json.name}>`;

  return str;
};

const getComponentName = (
  json: JSXLiteComponent,
  options: InternalToQootOptions,
) => {
  return capitalize(camelCase(json.name || 'my-component'));
};

// TODO
const getProvidersString = (
  componentJson: JSXLiteComponent,
  options: InternalToQootOptions,
): string => {
  return 'null';
};

const getEventHandlerFiles = (
  componentJson: JSXLiteComponent,
  options: InternalToQootOptions,
): File[] => {
  const files: File[] = [];

  traverse(componentJson).forEach(function (item) {
    if (isJsxLiteNode(item)) {
      for (const binding in item.bindings) {
        if (binding.startsWith('on')) {
          const componentName = getComponentName(componentJson, options);
          let str = dedent`
            import {
              injectEventHandler,
              provideQrlExp,
            } from 'qoot';
            import { ${componentName}Component } from './component'
            
            export default injectEventHandler(
              ${componentName}Component,
              provideQrlExp('event')
              async function (event) {
                ${removeSurroundingBlock(
                  processBinding(item.bindings[binding] as string, options),
                )}
                markDirty(this);
              }
            )
          `;

          str = format(str, {
            parser: 'typescript',
            plugins: [require('prettier/parser-typescript')],
          });
          files.push({
            path: `${componentName}/on${item.meta.id}${binding.slice(2)}.ts`,
            contents: str,
          });
        }
      }
    }
  });

  return files;
};

export type File = {
  path: string;
  contents: string;
};

export const componentToQoot = (
  componentJson: JSXLiteComponent,
  toQootOptions: ToQootOptions = {},
) => {
  let json = fastClone(componentJson);
  const options = {
    ...toQootOptions,
    namesMap: {},
    componentJson: json,
  };
  if (options.plugins) {
    json = runPreJsonPlugins(json, options.plugins);
  }
  const addWrapper = json.children.length > 1;
  if (options.plugins) {
    json = runPostJsonPlugins(json, options.plugins);
  }
  const componentName = capitalize(
    camelCase(componentJson.name || 'my-component'),
  );
  stripMetaProperties(json);
  let str = dedent`
    import { inject, QRL } from 'qoot';
    import { ${componentName}Component } from './component'
    ${renderPreComponent(json)}

    export default inject(${componentName}Component, function () {
      return (${addWrapper ? '<>' : ''}
        ${json.children.map((item) => blockToQoot(item, options)).join('\n')}
        ${addWrapper ? '</>' : ''})
    })
  `;

  if (options.plugins) {
    str = runPreCodePlugins(str, options.plugins);
  }
  if (options.prettier !== false) {
    str = format(str, {
      parser: 'typescript',
      plugins: [require('prettier/parser-typescript')],
    });
  }
  if (options.plugins) {
    str = runPostCodePlugins(str, options.plugins);
  }

  const dataString = getStateObjectString(json, {
    format: 'class',
    valueMapper: (code) => processBinding(code, options),
  });

  return {
    files: [
      {
        path: `${componentName}/template.tsx`,
        contents: str,
      },
      {
        path: `${componentName}/public.ts`,
        contents: dedent`
          import { jsxDeclareComponent, QRL } from 'qoot';
          export const ${componentName} = jsxDeclareComponent('${kebabCase(
          componentName,
        )}', QRL\`ui:/${componentName}/template\`);
        `,
      },
      {
        path: `${componentName}/component.ts`,
        contents: dedent`
          import { Component } from 'qoot';
          export class ${componentName}Component extends Component {
            ${dataString}
          }
        `,
      },
      ...getEventHandlerFiles(json, options),
    ],
  };
};
