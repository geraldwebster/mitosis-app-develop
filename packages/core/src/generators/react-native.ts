import json5 from 'json5';
import { camelCase, size } from 'lodash';
import { fastClone } from '../helpers/fast-clone';
import traverse from 'traverse';
import { ClassStyleMap } from '../helpers/collect-styles';
import { isJsxLiteNode } from '../helpers/is-jsx-lite-node';
import { Plugin } from '../modules/plugins';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { componentToReact } from './react';

type ToReactNativeOptions = {
  prettier?: boolean;
  stylesType?: 'emotion' | 'react-native';
  stateType?: 'useState' | 'mobx' | 'valtio' | 'solid' | 'builder';
  plugins?: Plugin[];
};

export const collectReactNativeStyles = (
  json: JSXLiteComponent,
): ClassStyleMap => {
  const styleMap: ClassStyleMap = {};

  const componentIndexes: { [className: string]: number | undefined } = {};

  traverse(json).forEach(function (item) {
    if (isJsxLiteNode(item)) {
      if (typeof item.bindings.css === 'string') {
        const value = json5.parse(item.bindings.css);
        delete item.bindings.css;
        if (!size(value)) {
          return;
        }
        const componentName = camelCase(item.name || 'view');
        const index = (componentIndexes[componentName] =
          (componentIndexes[componentName] || 0) + 1);
        const className = `${componentName}${index}`;
        item.bindings.style = `styles.${className}`;

        styleMap[className] = value;
      }
    }
  });

  return styleMap;
};

export const componentToReactNative = (
  componentJson: JSXLiteComponent,
  options: ToReactNativeOptions,
) => {
  const json = fastClone(componentJson);
  traverse(json).forEach((node) => {
    if (isJsxLiteNode(node)) {
      // TODO: handle TextInput, Image, etc
      if (node.name.toLowerCase() === node.name) {
        node.name = 'View';
      }

      if (
        node.properties._text?.trim().length ||
        node.bindings._text?.trim()?.length
      ) {
        node.name = 'Text';
      }
    }
  });

  return componentToReact(json, {
    ...options,
    stylesType: options.stylesType || 'react-native',
    type: 'native',
  });
};
