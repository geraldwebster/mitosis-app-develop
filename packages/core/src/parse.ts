import * as babel from '@babel/core';
import generate from '@babel/generator';
import JSON5 from 'json5';
import { JSXLiteNode } from './types/jsx-lite-node';
import { JSONObject, JSONOrNode, JSONOrNodeObject } from './types/json';
import { createJSXLiteNode } from './helpers/create-jsx-lite-node';
import { JSXLiteComponent, JSXLiteImport } from './types/jsx-lite-component';
import { createJSXLiteComponent } from './helpers/create-jsx-lite-component';

const jsxPlugin = require('@babel/plugin-syntax-jsx');
const tsPreset = require('@babel/preset-typescript');

export const selfClosingTags = new Set(['input', 'meta', 'bar']);

const { types } = babel;

type Context = {
  // Babel has other context
  builder: {
    component: JSXLiteComponent;
  };
};

const arrayToAst = (array: JSONOrNode[]) => {
  return types.arrayExpression(array.map((item) => jsonToAst(item)) as any);
};

const jsonToAst = (json: JSONOrNode): babel.Node => {
  if (types.isNode(json as any)) {
    if (types.isJSXText(json as any)) {
      return types.stringLiteral((json as any).value);
    }
    return json as babel.Node;
  }
  switch (typeof json) {
    case 'undefined':
      return types.identifier('undefined');
    case 'string':
      return types.stringLiteral(json);
    case 'number':
      return types.numericLiteral(json);
    case 'boolean':
      return types.booleanLiteral(json);
    case 'object':
      if (!json) {
        return types.nullLiteral();
      }
      if (Array.isArray(json)) {
        return arrayToAst(json);
      }
      return jsonObjectToAst(json as JSONObject);
  }
};

const jsonObjectToAst = (
  json: JSONOrNodeObject,
): babel.types.ObjectExpression => {
  if (!json) {
    return json;
  }
  const properties: babel.types.ObjectProperty[] = [];
  for (const key in json) {
    const value = json[key];
    if (value === undefined) {
      continue;
    }
    const keyAst = types.stringLiteral(key);
    const valueAst = jsonToAst(value);
    properties.push(types.objectProperty(keyAst, valueAst as any));
  }
  const newNode = types.objectExpression(properties);

  return newNode;
};

const componentFunctionToJson = (
  node: babel.types.FunctionDeclaration,
  context: Context,
): JSONOrNode => {
  let state = {};
  for (const item of node.body.body) {
    if (types.isVariableDeclaration(item)) {
      const init = item.declarations[0].init;
      if (types.isCallExpression(init)) {
        const firstArg = init.arguments[0];
        if (types.isObjectExpression(firstArg)) {
          const obj = JSON5.parse(generate(firstArg).code);
          state = obj;
        }
      }
    }
  }

  const theReturn = node.body.body.find((item) =>
    types.isReturnStatement(item),
  );
  const children: JSXLiteNode[] = [];
  if (theReturn) {
    const value = (theReturn as babel.types.ReturnStatement).argument;
    if (types.isJSXElement(value)) {
      children.push(jsxElementToJson(value) as JSXLiteNode);
    }
  }

  return {
    ...context.builder.component,
    state,
    children,
  } as any;
};

const jsxElementToJson = (
  node: babel.types.JSXElement | babel.types.JSXText,
): JSXLiteNode => {
  if (types.isJSXText(node)) {
    return createJSXLiteNode({
      properties: {
        _text: node.value,
      },
    });
  }
  return createJSXLiteNode({
    name: (node.openingElement.name as babel.types.JSXIdentifier).name,
    properties: node.openingElement.attributes.reduce((memo, item) => {
      if (types.isJSXAttribute(item)) {
        const key = item.name.name as string;
        const value = item.value;
        if (types.isStringLiteral(value)) {
          memo[key] = value;
          return memo;
        }
      }
      return memo;
    }, {} as { [key: string]: JSONOrNode }) as any,
    bindings: node.openingElement.attributes.reduce((memo, item) => {
      if (types.isJSXAttribute(item)) {
        const key = item.name.name as string;
        const value = item.value;

        if (types.isJSXExpressionContainer(value)) {
          const { expression } = value;
          if (types.isArrowFunctionExpression(expression)) {
            memo[key] = generate(expression.body).code;
          } else {
            memo[key] = generate(expression).code;
          }
          return memo;
        }
      } else if (types.isJSXSpreadAttribute(item)) {
        // TODO: potentially like Vue store bindings and properties as array of key value pairs
        // too so can do this accurately when order matters. Also tempting to not support spread,
        // as some frameworks do not support it (e.g. Angular) tho Angular may be the only one
        memo._spread = types.stringLiteral(generate(item.argument).code);
      }
      return memo;
    }, {} as { [key: string]: JSONOrNode }) as any,
    children: node.children.map((item) => jsxElementToJson(item as any)) as any,
  });
};

export function parse(jsx: string): JSXLiteComponent {
  const output = babel.transform(jsx, {
    presets: [[tsPreset, { isTSX: true, allExtensions: true }]],
    plugins: [
      jsxPlugin,
      () => ({
        visitor: {
          Program(path: babel.NodePath<babel.types.Program>, context: Context) {
            if (context.builder) {
              return;
            }
            context.builder = {
              component: createJSXLiteComponent(),
            };

            const isImportOrDefaultExport = (node: babel.Node) =>
              types.isExportDefaultDeclaration(node) ||
              types.isImportDeclaration(node);

            const keepStatements = path.node.body.filter((statement) =>
              isImportOrDefaultExport(statement),
            );
            const cutStatements = path.node.body.filter(
              (statement) => !isImportOrDefaultExport(statement),
            );

            // TODO: support multiple? e.g. for others to add imports?
            context.builder.component.hooks.preComponent = generate(
              types.program(cutStatements),
            ).code;

            path.replaceWith(types.program(keepStatements));
          },
          FunctionDeclaration(
            path: babel.NodePath<babel.types.FunctionDeclaration>,
            context: Context,
          ) {
            const { node } = path;
            if (types.isIdentifier(node.id)) {
              const name = node.id.name;
              if (name[0].toUpperCase() === name[0]) {
                path.replaceWith(
                  jsonToAst(componentFunctionToJson(node, context)),
                );
              }
            }
          },
          ImportDeclaration(
            path: babel.NodePath<babel.types.ImportDeclaration>,
            context: Context,
          ) {
            const importObject: JSXLiteImport = {
              imports: {},
              path: path.node.source.value,
            };
            for (const specifier of path.node.specifiers) {
              if (types.isImportSpecifier(specifier)) {
                importObject.imports[
                  (specifier.imported as babel.types.Identifier).name
                ] = specifier.local.name;
              } else if (types.isImportDefaultSpecifier(specifier)) {
                importObject.imports[specifier.local.name] = 'default';
              } else if (types.isImportNamespaceSpecifier(specifier)) {
                importObject.imports[specifier.local.name] = '*';
              }
            }
            context.builder.component.imports.push(importObject);
            // console.log('context 1', context);
            path.remove();
          },
          ExportDefaultDeclaration(
            path: babel.NodePath<babel.types.ExportDefaultDeclaration>,
          ) {
            path.replaceWith(path.node.declaration);
          },
          JSXText(path: babel.NodePath<babel.types.JSXText>) {
            path.replaceWith(types.stringLiteral(path.node.value));
          },
          JSXElement(path: babel.NodePath<babel.types.JSXElement>) {
            const { node } = path;
            path.replaceWith(jsonToAst(jsxElementToJson(node)));
          },
        },
      }),
    ],
  });

  try {
    return JSON5.parse(
      output!.code!.trim().replace(/^\({/, '{').replace(/}\);$/, '}'),
    );
  } catch (err) {
    console.error('Could not parse code', output && output.code);
    throw err;
  }
}
