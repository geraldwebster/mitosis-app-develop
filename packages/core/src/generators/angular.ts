import dedent from 'dedent';
import { format } from 'prettier';
import { renderPreComponent } from '../helpers/render-imports';
import { selfClosingTags } from '../parse';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';

type ToAngularOptions = {
  prettier?: boolean;
};
const blockToAngular = (json: JSXLiteNode, options: ToAngularOptions = {}) => {
  if (json.properties._text) {
    return json.properties._text;
  }

  let str = `<${json.name} `;
  for (const key in json.properties) {
    const value = json.properties[key];
    str += ` ${key}="${value}" `;
  }
  for (const key in json.bindings) {
    const value = json.bindings[key] as string;
    // TODO: proper babel transform to replace. Util for this
    const useValue = value.replace(/state\./g, '');

    if (key.startsWith('on')) {
      const event = key.replace('on', '').toLowerCase();
      str += ` (${event})="${useValue}" `;
    } else {
      str += ` [${key}]="${useValue}" `;
    }
  }
  if (selfClosingTags.has(json.name)) {
    return str + ' />';
  }
  str += '>';
  if (json.children) {
    str += json.children
      .map((item) => blockToAngular(item, options))
      .join('\n');
  }

  str += `</${json.name}>`;
  return str;
};

export const componentToAngular = (
  json: JSXLiteComponent,
  options: ToAngularOptions = {},
) => {
  let str = dedent`
    import { Component } from '@angular/core';
    ${renderPreComponent(json)}

    @Component({
      template: \`
        ${json.children.map((item) => blockToAngular(item)).join('\n')}
      \`
    })
    export default class MyComponent {
      ${Object.keys(json.state)
        .map((key) => ` ${key} = ${json.state[key]};`)
        .join('\n')}
    }
  `;

  if (options.prettier !== false) {
    str = format(str, { parser: 'typescript' });
  }
  return str;
};
