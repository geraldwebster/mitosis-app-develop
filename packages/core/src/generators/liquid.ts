import { format } from 'prettier';
import { selfClosingTags } from '../parse';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';

type ToLiquidOptions = {
  prettier?: boolean;
};
// TODO: spread support
const blockToLiquid = (json: JSXLiteNode, options: ToLiquidOptions = {}) => {
  if (json.properties._text) {
    return json.properties._text;
  }

  let str = `<${json.name} `;

  for (const key in json.properties) {
    const value = json.properties[key];
    str += ` ${key}="${value}" `;
  }

  for (const key in json.bindings) {
    const value = json.bindings[key] as string;
    // TODO: proper babel transform to replace. Util for this
    const useValue = value.replace(/state\./g, '');

    if (key.startsWith('on')) {
      // Do nothing
    } else {
      str += ` ${key}="{{${useValue}}}" `;
    }
  }
  if (selfClosingTags.has(json.name)) {
    return str + ' />';
  }
  str += '>';
  if (json.children) {
    str += json.children.map((item) => blockToLiquid(item, options)).join('\n');
  }

  str += `</${json.name}>`;
  return str;
};

export const componentToLiquid = (
  json: JSXLiteComponent,
  options: ToLiquidOptions = {},
) => {
  let str = json.children.map((item) => blockToLiquid(item)).join('\n');

  if (options.prettier !== false) {
    str = format(str, { parser: 'html' });
  }
  return str;
};
