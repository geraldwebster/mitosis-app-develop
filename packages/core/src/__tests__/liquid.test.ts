import { componentToLiquid } from '../generators/liquid';
import { parse } from '../parse';
import { basic } from './data/basic';
import { inputBlock } from './data/blocks/input';
import { submitButtonBlock } from './data/blocks/submitButton';

describe('Liquid', () => {
  test('Basic', () => {
    const json = parse(basic);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });

  test('Input block', () => {
    const json = parse(inputBlock);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });

  test('Submit button block', () => {
    const json = parse(submitButtonBlock);
    const output = componentToLiquid(json);
    expect(output).toMatchSnapshot();
  });
});
