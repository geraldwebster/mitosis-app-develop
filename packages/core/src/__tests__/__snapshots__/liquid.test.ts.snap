// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Liquid Basic 1`] = `
"<div>
  <input value=\\"{{name}}\\" />

  Hello! I can run in React, Vue, Solid, or Liquid!
</div>
"
`;

exports[`Liquid Input block 1`] = `
"<input
  placeholder=\\"{{placeholder}}\\"
  type=\\"{{type}}\\"
  name=\\"{{name}}\\"
  value=\\"{{value}}\\"
  defaultValue=\\"{{defaultValue}}\\"
  required=\\"{{required}}\\"
/>
"
`;

exports[`Liquid Submit button block 1`] = `
"<button type=\\"submit\\">{{undefined}}</button>
"
`;
